Check CachedResourceLoader's document WeakPtr in more places

brentfulgham opened the pull request:
#### 9632def06c0f39742c6b9b5af8ca7259b44583e8
<pre>
Check CachedResourceLoader&apos;s document WeakPtr in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=258167">https://bugs.webkit.org/show_bug.cgi?id=258167</a>
&lt;rdar://106964279&gt;

Reviewed by Chris Dumez.

The CachedResourceLoader class holds a WeakPtr to a Document, which may be
a nullptr at different moments in loading. We inconsistently check for
nullptr, which telemetry shows is triggering some crashes.

This patch provides more rigorous checks for nullptr to avoid these crashes.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::cachedResource const):
(WebCore::CachedResourceLoader::checkInsecureContent const):
(WebCore::CachedResourceLoader::allowedByContentSecurityPolicy const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::canRequestAfterRedirection const):
(WebCore::CachedResourceLoader::canRequestInContentDispositionAttachmentSandbox const):

Canonical link: <a href="https://commits.webkit.org/265259@main">https://commits.webkit.org/265259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d4ffd00f89a7d260608793f8c0644fda8d278aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12050 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10421 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12742 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10593 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12956 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10567 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11354 "2 flakes 2 failures") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12448 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8658 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9424 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16688 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9576 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12828 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10010 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9168 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/2501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13422 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -221,6 +221,7 @@ CachedResourceLoader::~CachedResourceLoader()
 {
     ASSERT(!resourceURL.isNull());
+    ASSERT(m_document);
     return cachedResource(MemoryCache::removeFragmentIdentifierIfNeeded(m_document->completeURL(resourceURL)));
```
If `m_document` could be `nullptr` at any time, we should be doing more than just asserting.

Would like to see something more like this (maybe even remove the `ASSERT(m_document);` statement:

```suggestion
    ASSERT(m_document);
    return m_document ? cachedResource(MemoryCache::removeFragmentIdentifierIfNeeded(m_document->completeURL(resourceURL))) : nullptr;
```

ddkilzer commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -447,10 +448,11 @@ bool CachedResourceLoader::checkInsecureContent(CachedResource::Type type, const
+            ASSERT(localTop->document());
             if (&top != frame && localTop && !MixedContentChecker::canRunInsecureContent(*localTop, localTop->document()->securityOrigin(), url))
                 return false;
         }
```
Why add a `nullptr` check in one case, but only an `ASSERT()` in the other case?

Also, the `ASSERT(localTop->document())` will crash if `localTop` is `nullptr`.

I'd like to see us be consistent here.

brentfulgham replied:
My assumption here was that if we are starting from a frame, we would have a document, but since it's a RefPtr I guess there are moments where that may not be true. I'll revise that.

ddkilzer commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -469,10 +471,11 @@ bool CachedResourceLoader::checkInsecureContent(CachedResource::Type type, const
+            ASSERT(localTopFrame->document());
             if (!localTopFrame || !MixedContentChecker::canDisplayInsecureContent(*localTopFrame, localTopFrame->document()->securityOrigin(), contentTypeFromResourceType(type), url))
                 return false;
         }
```
Ditto.

ddkilzer commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -498,44 +501,47 @@ bool CachedResourceLoader::allowedByContentSecurityPolicy(CachedResource::Type t
 
     ASSERT(m_document);
     ASSERT(m_document->contentSecurityPolicy());
+    auto contentSecurityPolicy = m_document->contentSecurityPolicy();
+    if (!contentSecurityPolicy)
+        return true;
```
This will still crash in Release/Production builds if `m_document` is `nullptr`.  We should add a `nullptr` check, not just assert.

This will still crash in Release/Production builds if `m_document` is `nullptr`.  We should add a `nullptr` check, not just assert.  Maybe:

```
    ASSERT(m_document);
    if (!m_document)
        return false;  // Or true???
```

brentfulgham replied:
I initially had that, but realized that all calls to this method already nullptr check document. But I can add it in case a new call site is added that does not do the check.

brentfulgham replied:
I think we return true; if no document, then no security origin, therefore no security boundary to check against, therefore no violation.

ddkilzer commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -747,6 +757,7 @@ bool CachedResourceLoader::canRequestInContentDispositionAttachmentSandbox(Cache
 
+    ASSERT(document);
     if (!document->shouldEnforceContentDispositionAttachmentSandbox() || document->securityOrigin().canRequest(url, OriginAccessPatternsForWebProcess::singleton()))
         return true;
```
Should also do a `document` `nullptr` check here, not just assert.

```suggestion
    ASSERT(document);
    if (document && (!document->shouldEnforceContentDispositionAttachmentSandbox() || document->securityOrigin().canRequest(url, OriginAccessPatternsForWebProcess::singleton())))
        return true;
```

ddkilzer replied:
Also need a `document` check in the code below (not part of the original patch):

```
    if (document) {
        String message = "Unsafe attempt to load URL " + url.stringCenterEllipsizedToLength() + " from document with Content-Disposition: attachment at URL " + document->url().stringCenterEllipsizedToLength() + ".";
        document->addConsoleMessage(MessageSource::Security, MessageLevel::Error, message);
    }
```

brentfulgham replied:
Will do. I think I'll check and early return, just like the early part of this method.

ddkilzer reviewed (changes requested):
I'm marking this "Request changes" as I'd like to see runtime `nullptr` checks if `m_document` is `WeakPtr<>` and could be `nullptr` at any time, and more consistent `nullptr` runtime (non-assert) checks.

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1682,9 +1707,14 @@ void CachedResourceLoader::warnUnusedPreloads()
     if (!m_preloads)
         return;
+
+    auto* document = this->document();
```
Here you cache a raw pointer...

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1682,9 +1707,14 @@ void CachedResourceLoader::warnUnusedPreloads()
         return;
+
+    auto* document = this->document();
+    if (!document)
```
And null check it here...

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1682,9 +1707,14 @@ void CachedResourceLoader::warnUnusedPreloads()
     for (const auto& resource : *m_preloads) {
-        if (resource.isLinkPreload() && resource.preloadResult() == CachedResource::PreloadResult::PreloadNotReferenced && document()) {
-            document()->addConsoleMessage(MessageSource::Other, MessageLevel::Warning,
+        if (resource.isLinkPreload() && resource.preloadResult() == CachedResource::PreloadResult::PreloadNotReferenced && document) {
```
... so the null check here for every resource does nothing.

Maybe you intended to store the document in a WeakPtr?

brentfulgham replied:
No -- I missed these cases. I can remove them. document() is already a week pointer, and there should be no script running in these calls to change that state.

cdumez reviewed (approved):
LGTM

brentfulgham committed:
Check CachedResourceLoader's document WeakPtr in more places
https://bugs.webkit.org/show_bug.cgi?id=258167
<rdar://106964279>

Reviewed by Chris Dumez.

The CachedResourceLoader class holds a WeakPtr to a Document, which may be
a nullptr at different moments in loading. We inconsistently check for
nullptr, which telemetry shows is triggering some crashes.

This patch provides more rigorous checks for nullptr to avoid these crashes.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::cachedResource const):
(WebCore::CachedResourceLoader::checkInsecureContent const):
(WebCore::CachedResourceLoader::allowedByContentSecurityPolicy const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::canRequestAfterRedirection const):
(WebCore::CachedResourceLoader::canRequestInContentDispositionAttachmentSandbox const):

Canonical link: https://commits.webkit.org/265259@main
